Added methods to define default sort

raphaeltraviss opened the pull request:
My first ever pull request!

OK, so with this pull, we can define default sorts and field limiters with defaultSortInfo() and defaultFieldInfo().  If the client passes a query string in the URL, they are completely overridden.

One thing I tried to do was to add the processing methods to check for invalid stuff, and allow you to supply the defaults in the same way as you specify them in the URL; e.g. instead of DESC/ASC you can simply supply 'fieldname' or '-fieldname'.

Field filters (as opposed to field limiters) seem to be more complicated, so I've leaving them for now.

So... how does it look?

raphaeltraviss committed:
Added methods to define default sort and limiter fields.

e0ipso commented on README.md:
```diff
@@ -202,6 +244,47 @@ array(
+Prefixing the sort field names with a minus (-) sign will apply a descending sort,
+instead of the default ascending.
+
+This default sort will be nullified if the request URL contains a sort query.
```
_This is only the default sort, any explicit sort declared in the URL will take precedence_

e0ipso commented on README.md:
```diff
@@ -143,6 +143,48 @@ array(
+should match a field defined in publicFieldsInfo().
+
+These field limiters will be nullified if the request URL contains a sort query.
+
```
I can only see a use case for this and I'm not sure it is worth adding another layer of complexity.

Use case I see is:
1. You declare a resource with lots of fields.
2. You limit the fields to only show a bunch of them.
3. Whenever you need any of the fields not being shown, you use the URL options to bring them.

Otherwise you can just remove the fields from `publicFieldsInfo`.

Not sure if this edge case is worth the trouble.

e0ipso replied:
@amitaibu thoughts? I'm leaning towards leaving this out, but I can be convinced otherwise. :arrow_up:

e0ipso commented on README.md:
```diff
@@ -202,6 +244,47 @@ array(
+    return $public_fields;
+  }
+
+  public function defaultSortInfo() {
```
I like this idea of allowing an optional default sort.

2 minor things:
- [ ] Make sure that the docs say that this is optional.
- [ ] Make sure to use a different example, since the default sort is based on entity id (even though you are using the reverse order).

e0ipso commented:
w00t! Thanks a lot for contributing @rfx1007!!!

amitaibu commented on README.md:
```diff
@@ -143,6 +143,48 @@ array(
+
+  // ...
+
+  public function publicFieldsInfo() {
```
this can be removed as the `\RestfulEntityBaseTaxonomyTerm` already provides ID and label.

amitaibu commented on README.md:
```diff
@@ -143,6 +143,48 @@ array(
+    return $public_fields;
+  }
+
+  public function defaultSortInfo() {
```
add PHPdocs even in the docs.

amitaibu commented on README.md:
```diff
@@ -143,6 +143,48 @@ array(
+  }
+
+  public function defaultSortInfo() {
+    return array('-tid');
```
this should be 

``` php
return array(
  'id' => 'DESC',
);
```

amitaibu commented on README.md:
````diff
@@ -202,6 +244,47 @@ array(
+This default sort will be nullified if the request URL contains a sort query.
+
+```php
+class MyPlugin extends \RestfulEntityBaseTaxonomyTerm {
````
seems the docs are repeating -- it's enough to describe once `defaultSortInfo`

amitaibu commented on plugins/restful/RestfulEntityBase.php:
```diff
@@ -281,6 +281,10 @@ public function viewEntity($entity_id) {
 
     $limit_fields = !empty($request['fields']) ? explode(',', $request['fields']) : array();
 
+    if (empty($limit_fields)) {
```
remove

raphaeltraviss replied:
The purpose of this empty() check is to add in the default field limiters if there are none supplied via the URL.  Are you saying that we should remove the empty() check, and have the default field limiters added all the time, even if some are supplied via the URL?  This would make it impossible to fully override the default field limiters... am I misinterpreting your 'remove' comment?

amitaibu replied:
Yes, sorry - my intention was that that the entire field limiters functionality should be removed. If you wish to provide those limiters you may do it via `access callbacks` on each public field.

raphaeltraviss replied:
Ok, so the intent is to remove all field limiter functionality altogether, including the ?fields capability from the URL, and therefore, we should remove all field limiter functionality from this pull request?

amitaibu replied:
Indeed, the PR should be only about defaultSort (I'm sorry if I misguided you in my previous comments about those limiters),

amitaibu commented on plugins/restful/RestfulEntityBase.php:
```diff
@@ -1175,6 +1179,94 @@ public function getPublicFields() {
+   *
+   * @throws Exception
+   */
+  protected function processDefaultSorts() {
```
I don't think we need it, we could pas the default sort to our regular sort -- and if there's an error there it will throw the exception.

raphaeltraviss replied:
These methods handle invalid values that could be supplied in the defaultXInfo methods, which I've thought of a couple ways to deal with:
1. provide simplified copies of the parseRequest methods as processX methods.  Then, the data provider classes would need to call processX methods in addition to parseRequestForX methods.
2. change the name of parseRequestForX methods and modify them to merge values from both the request and the defaultXInfo methods.  The data provider classes would need to call the method by its new name.
3. Don't handle invalid values supplied via the plugin definition at all; let them hit the sorting/limiting logic and throw an exception there.  The data provider classes would call the defaultXInfo() methods directly.

I was leaning towards options 1 or 3, but I decided to try option 1, because I felt that having the processX methods was more consistent with the existing pattern of calling parseRequestForX in the data provider, and then having parseRequest sanitize the options.

On the other hand, I can see where it's not technically necessary to have this consistency, since the plugin-supplied options are being supplied by the plugin developer him/herself, and not by an API consumer.

So, should I go ahead with option 3?

amitaibu replied:
> So, should I go ahead with option 3?

Yes, simply because it's the simplest solution.

raphaeltraviss committed:
Removed field limiter logic and simplified sorting logic.

raphaeltraviss committed:
Moved defaultSortInfo() to RestfulBase.

raphaeltraviss commented:
Ok, I pared it down a bit.

Also, I moved defaultSortInfo() to RestfulBase because it should work for both EFQ and DbQuery plugins; it's working as expected with my custom query plugin.

amitaibu commented on README.md:
````diff
@@ -202,6 +202,47 @@ array(
 ```
 
+You can also define default sort fields in your plugin, by overriding
+defaultSortInfo() in your class definition.
````
wrap `defaultSortInfo()` and `publicFieldsInfo()`  in <code>``</code>

amitaibu commented on README.md:
```diff
@@ -202,6 +202,47 @@ array(
+This method should return an associative array, with each element having a key
+that matches a field from publicFieldsInfo(), and a value of either 'ASC' or 'DESC'.
+
+This default sort will be nullified if the request URL contains a sort query.
```
will be nullified => will be ignored

amitaibu commented on README.md:
```diff
@@ -202,6 +202,47 @@ array(
+  /**
+   * Overrides \RestfulEntityBase::publicFieldsInfo().
+   */
+  public function publicFieldsInfo() {
```
I don't think `publicFieldsInfo()` is needed here, enough to show defaultSortInfo()

amitaibu commented on README.md:
```diff
@@ -202,6 +202,47 @@ array(
+   */
+  public function defaultSortInfo() {
+    // Sort by 'field_myfield' in descending order.
+    return array('myfield' => 'DESC');
```
Maybe keep it basic?
`return array('id' => 'DESC');`

e0ipso replied:
I actually like the idea of keeping `myfield` following the declared field above.

amitaibu commented on plugins/restful/RestfulBase.php:
```diff
@@ -882,6 +882,17 @@ public function getPublicFields() {
   /**
+   * Defines default sort fields if none are provided via the request URL.
+   *
+   * @return array of key/value pairs:
```
``` php
* @return array
*    Array keyed by the public field name, and the order ('ASC' or 'DESC') as value.
```

amitaibu commented on plugins/restful/RestfulDataProviderDbQuery.php:
```diff
@@ -121,8 +121,14 @@ public function getQueryForList() {
     $sorts = $this->parseRequestForListSort();
+
+    if (empty($sorts)) {
+      $sorts = $this->defaultSortInfo();
```
`$sorts = $sorts ? $sorts : $this->defaultSortInfo();`

amitaibu commented:
PR is getting closer :)
Now we also need to make sure it's tested in `\RestfulListTestCase::testSort()`

You can add for example to `RestfulTestArticlesResource__1_1` a `defaultSortInfo()` and then test it:

``` php
$handler = restful_get_restful_handler('test_articles', 1, 1);
$result = $handler->get();

// Assert IDs are sorted descending.
```

amitaibu commented on plugins/restful/RestfulBase.php:
```diff
@@ -882,6 +882,17 @@ public function getPublicFields() {
+   *   value must be one of either ASC or DESC.
+   */
+  public function defaultSortInfo() {
+    return array();
```
I think this should actually be populated with something like:

``` php
  // Some endpoints like 'token_auth' don't have an id public field. In that
  // case, skip the default sorting.
  return !empty($public_fields['id']) ? array('id' => 'ASC') : array();
```

and for DB: `return array($this->getIdColumn() => 'ASC')`

So we can avoid that logic in `queryForListSort`

e0ipso replied:
:+1:

raphaeltraviss committed:
Merge branch '7.x-1.x' into default-sort-filter

raphaeltraviss committed:
Moved defaultSortInfo() back into dataProvider logic; doc fixes.

e0ipso commented on plugins/restful/RestfulDataProviderEFQ.php:
```diff
@@ -95,6 +95,11 @@ public function getQueryForList() {
     $sorts = $this->parseRequestForListSort();
+
+    if (empty($sorts)) {
+      $sorts = $this->defaultSortInfo();
```
`$sorts = $sorts ? $sorts : $this->defaultSortInfo();`

e0ipso commented:
@rfx1007 please, let me know if you need more clarification. I understand that you are not as familiar with the codebase as we are :-)

raphaeltraviss committed:
Added tests for defaultSortInfo().

raphaeltraviss commented:
Moved the logic back into the data providers, since they have different defaults.  Also, added a test for defaultSortInfo() and a test for overriding it.

raphaeltraviss committed:
Merge branch '7.x-1.x' into default-sort-filter

raphaeltraviss committed:
Only add default sort if default field is present.

amitaibu commented:
@rfx1007 you are getting very close, just need to get those tests back to green! Don't be discouraged - everybody hates (fixing the) tests ;)

raphaeltraviss committed:
Merge branch '7.x-1.x' into default-sort-filter

raphaeltraviss committed:
Updated sort position on render cache test.

raphaeltraviss committed:
Fake default sort in render cache test.

e0ipso commented on tests/RestfulRenderCacheTestCase.test:
```diff
@@ -34,6 +34,14 @@ class RestfulRenderCacheTestCase extends DrupalWebTestCase {
+    function labelSort($a, $b) {
+      return strcasecmp($a->title, $b->title);
+    }
+    usort($nodes, 'labelSort');
```
I don't understand why we need this. Was this causing test failures? Is so, why?

raphaeltraviss replied:
The failures were caused by the random title on the test nodes.  Since we are sorting by label, this was causing the nodes to be sorted in a different order than they were originally created in.

Consider:

``` php
    $node = reset($nodes);
    $cid = 'v1.1' . '::uu' . $account->uid . '::pa' . 'et:node::ei:' . $node->nid;
    $record = $cache->get($cid);
    $this->assertEqual($pre_cache_results[0], $record->data, 'Data in cache bins is correct.');
```

$pre_cache_results is being built from $handler->get(), which is now being sorted by label, whereas $record->data is being built from the nid of the first element of $nodes, which is ordered by id, so the test will fail 66% of the time, depending on what the random titles generate.

By applying the sort function to $nodes, reset($nodes) and the $pre_cache_results[0] now point to the same node, so the test is now valid.

amitaibu replied:
I agree with @mateu-aguilo-bosch - lets avoid complex non-standard stuff in the tests. If the difficulty is the random titles - then go ahead and use hardcoded ones.

raphaeltraviss replied:
hard-coded the titles with $index so they are sorted in the expected order.

raphaeltraviss committed:
Simplified render cache tests.

amitaibu commented on tests/RestfulRenderCacheTestCase.test:
```diff
@@ -29,7 +29,7 @@ class RestfulRenderCacheTestCase extends DrupalWebTestCase {
     $num_articles = 3;
     for ($index = 0; $index < $num_articles; $index++) {
-      $settings['title'] = $this->randomName();
+      $settings['title'] = 'Node title '.$index;
```
`'Node title '.$index` => `'Node title ' . $index`

amitaibu commented on tests/RestfulRenderCacheTestCase.test:
```diff
@@ -79,7 +80,9 @@ class RestfulRenderCacheTestCase extends DrupalWebTestCase {
     $this->assertFalse($record, 'Cache bin is emptied after updating a user.');
 
-    // Test cache segmentation with request params.
+    $node = reset($nodes);
```
can you add a comment about this part?

raphaeltraviss replied:
I just set this back to the way it was; now that the hard-coded titles sort in the expected order, we don't need to reset $node and $cid after appending to one.

raphaeltraviss committed:
More render cache test fixes.

amitaibu commented:
Well done, merged!

raphaeltraviss commented:
:trophy: finally legit!

amitaibu commented:
@rfx1007 I don't know what your twitter name is, but you should defiantly brag about this ;)

e0ipso commented:
YES!
